DPGAnalysis/SiStripTools/plugins/APVCyclePhaseProducerFromL1TS.cc

```diff
@@ -185,38 +190,43 @@ void APVCyclePhaseProducerFromL1TS::produce(edm::Event& iEvent, const edm::Event
 
   Handle<Level1TriggerScalersCollection> l1ts;
   iEvent.getByToken(_l1tscollectionToken, l1ts);
-
+  Handle<TCDSRecord> tcds_pIn;
+  iEvent.getByToken(_tcdsRecordToken, tcds_pIn);
+  bool useTCDS(tcds_pIn.isValid() && !_forceSCAL);
+  const auto* tcdsRecord = useTCDS ? tcds_pIn.product() : nullptr;
   // offset computation
 
   long long orbitoffset = 0;
 
-  if (!l1ts->empty()) {
-    if ((*l1ts)[0].lastResync() != 0) {
-      orbitoffset = _useEC0 ? (*l1ts)[0].lastEventCounter0() + _magicOffset : (*l1ts)[0].lastResync() + _magicOffset;
+  if (useTCDS) {
+    // l1ts->empty() always retuns false (last commit as of today: https://github.com/cms-sw/cmssw/commit/f4694d795d4b268d541c633dfb68283d889264b0 ), so the check is likely not necessary---and TCDSRecord hasn't anything similar
+    if (tcdsRecord->getLastResync() != 0) {
```

Comment:
@vischia @mommsen 
I investigated a bit the error reported by the bot [here](https://github.com/cms-sw/cmssw/pull/29116#issuecomment-601833108) and I found out that this is the line triggering the problem reported:
```
----- Begin Fatal Exception 23-Mar-2020 11:45:26 CET-----------------------
An exception of category 'StdException' occurred while
   [0] Processing  Event run: 1 lumi: 1 event: 1 stream: 0
   [1] Prefetching for module SiStripMonitorCluster/'HLTSiStripMonitorCluster'
   [2] Calling method for module APVCyclePhaseProducerFromL1TS/'APVPhases'
Exception Message:
A std::exception was thrown.
vector::_M_range_check: __n (which is 5) >= this->size() (which is 0)
----- End Fatal Exception -------------------------------------------------
```

As far as I can see `getLastResync()` is implemented here:

https://github.com/cms-sw/cmssw/blob/1711d0fe46c07e72b3a77242f61df6fe98860fad/DataFormats/TCDS/interface/TCDSRecord.h#L121

and indeed `BGo::Resync` is the fifth element of this enum:

https://github.com/cms-sw/cmssw/blob/1711d0fe46c07e72b3a77242f61df6fe98860fad/DataFormats/TCDS/interface/TCDSRecord.h#L27

It seems to me that the `lastBgos_` vector is empty.
The two failing workflows (`11634.0` and `12434.0`) are both Run3 simulation workflows, and therefore in both cases  `forceSCAL` is set to `False`. 
What should we expect for the TCDS unpacker in the simulation case? Is this vector supposed to be filled?

Reply:
If the issue is with simulation would suggest to adopt the same strategy adopted for SCAL. Also in that case no one expected that the simulation provided sensible values.

Reply:
@venturia I tested this 
```suggestion
    if (tcdsRecord->getLastResync() != 0 && iEvent.eventAuxiliary().isRealData()) {
```
and I confirm it works (at least it does not crash). Though I would like the DAQ team to confirm that's the expected behaviour.

Reply:
There's indeed nothing sensible to be simulated for the TCDS payload. Thus, you should rely upon TCDS information only for real data.

Reply:
Thank you @mommsen 
@vischia can you implement what has been suggested above?

Reply:
@mmusich @mommsen @davidlange6 Done, but in the version suggested by @mommsen 

```
 if (useTCDS && iEvent.eventAuxiliary().isRealData()) {
```
because it makes sense to me that the check is performed for all the conditionals inside that if.

@davidlange6 was suggesting to add the condition only to the inner if in https://github.com/cms-sw/cmssw/pull/29116#issuecomment-602562860

```
if (tcdsRecord->getLastResync() != 0 && iEvent.eventAuxiliary().isRealData()) {
```
and reports it does not crash, so maybe on the other side it is not necessary to have the condition in the outer loop. Please let me know if that's the case (I implemented the one suggested by @mommsen , am pushing now)

Reply:
@vischia, David was just copying the earlier version of my own suggestion, which was buggy. 
Please go ahead with the solution I suggested here: https://github.com/cms-sw/cmssw/pull/29116#discussion_r396394638

Reply:
@mmusich OK perfect, that's what I uploaded indeed. Thanks for confirming!